{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imputed gene expression \n",
    "## Introduction\n",
    "\n",
    "In this notebook, there are several sections that describe the refactored function.\n",
    "- GenotypeDataset function\n",
    "    - Revised full function\n",
    "    - Decomposing code for testing\n",
    "      + get_all_dosages\n",
    "      + get_reference\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libaraies\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sqlite3\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "from pathlib import Path\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### <span style=\"color:orange\">REFACTORED Here we refactored the function GenotypeDataset into the `imputed_predixcan.py`  modules.</span>\n",
    "+ <span style=\"color:orange\">REFACTORED</span> Revised full function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class WeightsDB():\n",
    "    \n",
    "    def __init__(self,\n",
    "                 beta_file):\n",
    "        self.conn = sqlite3.connect(beta_file)\n",
    "\n",
    "    def query(self,\n",
    "              sql,\n",
    "              args=None):\n",
    "\n",
    "        c = self.conn.cursor()\n",
    "        if args:\n",
    "            for ret in c.execute(sql, args):\n",
    "                yield ret\n",
    "        else:\n",
    "            for ret in c.execute(sql):\n",
    "                yield ret\n",
    "\n",
    "\n",
    "class GenotypeDataset():\n",
    "\n",
    "    @staticmethod\n",
    "    def get_all_dosages(genoytpe_dir,\n",
    "                        dosage_prefix,\n",
    "                        dosage_end_prefix,\n",
    "                        unique_rsids,\n",
    "                        reference_file):\n",
    "\n",
    "        for chrfile in [\n",
    "                x for x in sorted(genoytpe_dir.iterdir())\n",
    "                if x.name.startswith(str(dosage_prefix))\n",
    "                and x.name.endswith(str(dosage_end_prefix))\n",
    "        ]:\n",
    "            # get chr number\n",
    "            chr_name = os.path.basename(chrfile).split(\".\")[0]\n",
    "            chr_number = re.findall(r'\\d+', chr_name)\n",
    "            # get reference file with specific chr numner\n",
    "            get_ref_chrfile = reference_file.loc[reference_file.CHR == chr_number[0]]\n",
    "            get_ref_chrfile = get_ref_chrfile.loc[get_ref_chrfile['SNP'].isin(unique_rsids)]\n",
    "            # create chr_bp column\n",
    "            get_ref_chrfile['chr_bp'] = get_ref_chrfile.apply(lambda x: str(x['CHR']) + \":\" + str(x['BP']), axis=1)\n",
    "            get_ref_chrfile.drop_duplicates([\"chr_bp\"], inplace=True)\n",
    "\n",
    "            with open(str(chrfile), 'rt') as file:\n",
    "                for line_index, line in enumerate(file):\n",
    "\n",
    "                    if line_index <= 0:\n",
    "                        continue\n",
    "\n",
    "                    arr = line.strip().split()\n",
    "                    chr_bp = arr[0]\n",
    "                    refallele = arr[2]\n",
    "                    dosage_row = np.array(arr[3:], dtype=np.float64)\n",
    "\n",
    "                    get_rsid = get_ref_chrfile['chr_bp'].astype(str).isin([str(chr_bp)])\n",
    "                    if any(get_rsid):\n",
    "                        rsid = get_ref_chrfile.loc[get_rsid]['SNP'].tolist()[0]\n",
    "                        yield rsid, refallele, dosage_row\n",
    "                    else:\n",
    "                        continue\n",
    "\n",
    "    @staticmethod\n",
    "    def UniqueRsid(beta_file):\n",
    "        res = [\n",
    "            x[0] for x in WeightsDB(beta_file).query(\n",
    "                \"SELECT distinct rsid FROM weights\")\n",
    "        ]\n",
    "        return res\n",
    "\n",
    "    @staticmethod\n",
    "    def get_reference(file, \n",
    "                      chunkSize=1000000, \n",
    "                      parition=661):\n",
    "        reader = pd.read_csv(file, sep=\"\\t\", iterator=True)\n",
    "        chunks = []\n",
    "        with tqdm(range(parition)) as pbar:\n",
    "            for _ in pbar:\n",
    "                try:\n",
    "                    chunk = reader.get_chunk(chunkSize)\n",
    "                    chunks.append(chunk)\n",
    "                except StopIteration:\n",
    "                    break\n",
    "        return  pd.concat(chunks, ignore_index=True)\n",
    "\n",
    "\n",
    "beta_file = \"/exeh_4/yuping/Epistasis_Interaction/data/models_all/gtex_v7_Cells_Transformed_fibroblasts_imputed_europeans_tw_0.5_signif.db\"\n",
    "unique_rsids = GenotypeDataset.UniqueRsid(beta_file)\n",
    "reference_file = GenotypeDataset.get_reference(\"/mnt/data/share/yuping/snp151_GRCh37.txt\")\n",
    "genotype_dir = Path(\"/mnt/data/share/yuping/dosage\")\n",
    "dosage_prefix = \"chr\"\n",
    "dosage_end_prefix = \".dosage\"\n",
    "\n",
    "for rsid, allele, dosage_row in GenotypeDataset.get_all_dosages(genotype_dir, \n",
    "                                                                dosage_prefix,\n",
    "                                                                dosage_end_prefix,\n",
    "                                                                unique_rsids,\n",
    "                                                                reference_file):\n",
    "    print(rsid)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "+ <span style=\"color:orange\">REFACTORED</span> Decomposing code for testing - `get_all_dosages`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "beta_file = \"/mnt/data/share/yuping/data/models_all/gtex_v7_Cells_Transformed_fibroblasts_imputed_europeans_tw_0.5_signif.db\"\n",
    "unqiue_rsids = GenotypeDataset.UniqueRsid(beta_file)\n",
    "genotype_dir = Path(\"/mnt/data/share/yuping/dosage\")\n",
    "dosage_prefix = \"chr\"\n",
    "dosage_end_prefix = \".dosage\"\n",
    "\n",
    "for chrfile in [\n",
    "        x for x in sorted(genotype_dir.iterdir())\n",
    "        if x.name.startswith(str(dosage_prefix))\n",
    "        and x.name.endswith(str(dosage_end_prefix))\n",
    "]:\n",
    "\n",
    "    chr_name = os.path.basename(chrfile).split(\".\")[0]\n",
    "    chr_number = re.findall(r'\\d+', chr_name)\n",
    "    ref_file = reference_file.loc[reference_file.CHR == chr_number[0]]\n",
    "    ref_file = ref_file.loc[ref_file['SNP'].isin(unqiue_rsids)]\n",
    "\n",
    "    ref_file['chr_bp'] = ref_file.apply(\n",
    "        lambda x: str(x['CHR']) + \":\" + str(x['BP']), axis=1)\n",
    "    ref_file.drop_duplicates([\"chr_bp\"], inplace=True)\n",
    "\n",
    "    print(chr_number)\n",
    "\n",
    "    with open(str(chrfile), \"rt\") as file:\n",
    "        for line_index, line in enumerate(file):\n",
    "\n",
    "            if line_index <= 0:\n",
    "                continue\n",
    "\n",
    "            arr = line.strip().split()\n",
    "            chr_bp = arr[0]\n",
    "            refallele = arr[2]\n",
    "            dosage_row = np.array(arr[3:], dtype=np.float64)\n",
    "\n",
    "            get_rsid = ref_file['chr_bp'].astype(str).isin([str(chr_bp)])\n",
    "            if any(get_rsid):\n",
    "                rsid = ref_file.loc[get_rsid]['SNP'].tolist()[0]\n",
    "            else:\n",
    "                continue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "+ <span style=\"color:orange\">REFACTORED</span> Decomposing code for testing - `get_reference`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reference(file, chunkSize=1000000, parition=661):\n",
    "    reader = pd.read_csv(file, sep = \"\\t\", iterator=True)\n",
    "    chunks = []\n",
    "    with tqdm(range(parition)) as pbar:\n",
    "        for _ in pbar:\n",
    "            try:\n",
    "                chunk = reader.get_chunk(chunkSize)\n",
    "                chunks.append(chunk)\n",
    "            except StopIteration:\n",
    "                break\n",
    "\n",
    "    return pd.concat(chunks,ignore_index=True)\n",
    "\n",
    "file = get_reference(\"/mnt/data/share/yuping/snp151_GRCh37.txt\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "yp_r",
   "language": "python",
   "name": "yp_r"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
